pkg/labels/parse.go

```diff
@@ -122,21 +126,55 @@ func ParseMatcher(s string) (_ *Matcher, err error) {
 	}
 
 	var (
-		rawValue            = ms[3]
-		value               strings.Builder
-		escaped             bool
-		expectTrailingQuote bool
+		rawName  = ms[1]
+		rawValue = ms[3]
 	)
 
-	if strings.HasPrefix(rawValue, "\"") {
-		rawValue = strings.TrimPrefix(rawValue, "\"")
-		expectTrailingQuote = true
+	matcherType, ok := typeMap[ms[2]]
+	if !ok {
+		return nil, errors.Errorf("matcher type is unknown: %s", ms[2])
 	}
 
+	if !utf8.ValidString(rawName) {
+		return nil, errors.Errorf("matcher label not valid UTF-8: %s", rawName)
+	}
 	if !utf8.ValidString(rawValue) {
-		return nil, errors.Errorf("matcher value not valid UTF-8: %s", ms[3])
+		return nil, errors.Errorf("matcher value not valid UTF-8: %s", rawValue)
+	}
+
+	value, err := unescapeMatcherString(rawValue)
+	if err != nil {
+		return nil, fmt.Errorf("matcher value %w", err)
+	}
+
+	name := rawName
+	// if name is quoted, then it can contain any UTF-8 character. Unescape some escape sequences.
+	if strings.HasPrefix(rawName, `"`) {
```

Comment:
I thought we want to escape open metrics for unquoted strings, not for double quoted strings? Double quoted strings have different escape sequences right?

```suggestion
	if !strings.HasPrefix(rawName, `"`) {
```

Reply:
Unquoted names can be only Prometheus-compatible, and therefore no escaping is allowed. Double quoted in contrast, can contain any UTF-8 characters, and therefore it can be escaped the same way as value.